### PR TITLE
[gitlab] Upload unittest JUnit files to CI App

### PR DESCRIPTION
**Pull request by bkabrda**
### What does this PR do?

This PR ensures that unittest JUnit files from all `source_test` jobs are uploaded to the CI App.

* Each `source_test` job (that runs tests that produce JUnit) creates a tarball with multiple JUnit XML files - there are multiple because we have multiple Go modules, so there's e.g. the root `junit-out.xml` as well as `pkg/otlp/model/junit-out.xml`. This tarball also contains 2 special files, `job_url.txt` and `tags.txt`. These contain additional information that we submit to the CI App so the individual tests have proper metadata (see below).
* The `source_test_junit_upload` job takes all of the tarballs and extracts them one by one and submits all JUnit files that it finds in them:
  * Each JUnit file is split into multiple per-owner files, so the owner is set properly in the CI App.
  * Metadata from `job_url.txt` and `tags.txt` are used for `datadog-ci` invocation, to set correct metadata about os/architecture/job name/job url (otherwise metadata of the `source_test_junit_upload` would be attached).

TODO before merging:
* Uncomment the lines in `source_test_junit_upload` that would only make it run on `main` and stable branches.
* After https://github.com/DataDog/datadog-agent-buildimages/pull/181 is merged, use the proper buildimages tag in .gitlab-ci.yml.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: tasks/libs/junit_upload.py
Comment (KSerrania) [inline tasks/libs/junit_upload.py:31]: If we're tagging only by the first codeowner, we may want to rework some lines in the `CODEOWNERS` file, eg. https://github.com/DataDog/datadog-agent/blob/b377f242267dd3fe361cb6f99cbcc344144f518e/.github/CODEOWNERS#L159-L160, https://github.com/DataDog/datadog-agent/blob/b377f242267dd3fe361cb6f99cbcc344144f518e/.github/CODEOWNERS#L83, etc. where the "first" owner isn't necessarily the main owner.

File: .gitlab/source_test_junit_upload.yml
Comment (KSerrania) [inline .gitlab/source_test_junit_upload.yml:3]: Should we uncomment these lines, or is it fine to run this on all branches?

**Review by KSerrania — APPROVED** [review]
LGTM

File: .gitlab/source_test_junit_upload.yml
Comment (bkabrda) [reply .gitlab/source_test_junit_upload.yml:3]: I'm leaving these commented until I do all changes that I've wanted, because otherwise the job won't actually run and I won't be able to test the changes :) I have it on my todo list to uncomment before I merge.

File: tasks/libs/junit_upload.py
Comment (bkabrda) [reply tasks/libs/junit_upload.py:31]: Yeah, but this should really be decided by the individuals/teams that work on these, so I don't want to mix this in the PR. I'll send the communication about this when this is merged, so people can update codeowners as they want.

File: .gitlab/source_test_junit_upload.yml
Comment (KSerrania) [reply .gitlab/source_test_junit_upload.yml:3]: I'm not sure what kind of load the CI App supports, but it may be fine to leave this on in all branches?
Having this on all branches may make it easier to detect flaky tests (due to the volume of tests run vs. only reporting on main and release branches), and you can always filter by branch in the UI to only see the state of `main` + release branch tests if needed.

Comment (bkabrda) [reply .gitlab/source_test_junit_upload.yml:3]: TBH I don't feel like the data from branches can be used for anything, since the fact that a test fails on a feature branch doesn't tell you anything - you don't know if it was broken (and then fixed before merging) or if it flaked out. That's why I propose gathering the data from where we're sure it will be meaningful. We can always change this in the future.

Comment (albertvaka) [reply .gitlab/source_test_junit_upload.yml:3]: Let's discuss this in the daily today. I can see how having this on all branches would add "noise" in form of failed tests that were actually intended to fail, but on the other hand it would add so many more runs of all the other tests that actual flakes would stand out regardless.

Comment (albertvaka) [reply .gitlab/source_test_junit_upload.yml:3]: Also, if this is the only thing to discuss from the PR, we can probably merge it now before the code freeze and change that at any time later.

Comment (bkabrda) [reply .gitlab/source_test_junit_upload.yml:4]: Ok, after some more discussion and thinking, I guess we'll never know if there are useful data in feature branches unless we try - so I'll leave this commented for now.

**Pull request merged**
